portfolio/src/main/java/com/google/sps/servlets/ThumbsDownServlet.java

```diff
@@ -147,7 +151,7 @@ private void incrementPopularity(Entity commentEntity) {
             commentEntity.setProperty("popularity", newPopularity);
 
             // Add the updated entity back in the datastore
-            datastore.put(commentEntity);
+            datastore.put(txn, commentEntity);
             txn.commit();
         } finally {
             if (txn.isActive()) {
```

Comment:
You also need to use a transaction for the `addToUnlikedComments` and `removeFromUnlikedComments` methods (and the same with the corresponding methods in the ThumbsUpServlet). A good way to spot these race conditions is to look for the read-modify-store pattern. Anytime you do these three things in succession with a database record there is the potential for a race condition and so you need to put this in a transaction.

This might take some refactoring since you get the UserInfoEntity in a separate method and share it across methods. Ideally, you should get the latest UserInfoEntity each time you need to make a change (and do this inside a transaction).

Reply:
I see, I will attend to those right now then.

Reply:
Hi Anand! I just submitted changes to address the newest issues you brought up. This involved some changes to the way I was dealing with  the comment class and the user info class. I think this made things more organized but I am not sure if it tackled the issue of getting UserInfoEntity each time and using a transaction to do that. Do you think these changes help me in that direction?